reps/2023-06-30-serve-cpp-deployment.md

```diff
@@ -0,0 +1,411 @@
+Next, we start the Ray Serve runtime and use Python Serve API deploy these Service as Deployment: 
+```python
+feature_service = serve.deployment(_func_or_class='FeatureService::FactoryCreate', name='feature_service', language='CPP')
+similarity_service = serve.deployment(_func_or_class='SimilarityService::FactoryCreate', name='similarity_service', language='CPP')
+rank_service = serve.deployment(_func_or_class='RankService::FactoryCreate', name='rank_service', language='CPP')
+
+with InputNode() as input:
+    features = feature_service.GetVector.bind(input[0])
+    similarities = similarity_service.GetSimilarity.bind(features)
+    rank_result = rank_service.Rank.bind(similarities, input[1])
+
+graph = DAGDriver.bind(rank_result, http_adapter=json_request)
+handle = serve.run(graph)
+ref = handle.GetVector.remote()
+result = ray.get(ref)
+print(result)
+```
+### Calling Ray Serve Deployment with HTTP
+```java
+curl http://127.0.0.1:8000?request=computer&num=1
+```
+### Overall Design
+   ![image](https://user-images.githubusercontent.com/40862361/250100276-5eeed3a1-70a2-4208-8563-e0b1bf7f7fd2.png)  
+Ray Serve maintains Controller Actor and Ingress Actor. So these two roles are not related to the language of the user's choice. And they have the ability to manage cross-language deployments and route requests.
+### C++ Case Deduction
+The businesses can send control commands to the Controller Actor of Ray Serve, which includes creating ingress, creating deployment, etc. When publishing C++ online services, the DeploymentState component needs to create C++ Deployment  Actors. Users can call their business logic in Python/Java Driver, Ray task, or Ray Actor, and the requests will be dispatched to C++ Deployment Actors.
+![image](https://user-images.githubusercontent.com/40862361/250100685-96466519-1f8e-414e-8d8a-1bcb63952a41.png)
+### Package
+C++ programs are typically compiled and packaged into three types of results: binary, static library, shared library.  
+- binary: C++ does not have a standard load binary API, and binaries compiled by different compilers may not be the same. Directly loading the binary can result in many uncontrollable factors.   
+- static library: Static library bundle all their dependencies together, whereas common utility libraries such as glog and boost may be loaded via dynamic dependencies in the Ray Serve Deployment. This can lead to a higher probability of conflicts when the same library exists as both a dynamic and static dependency.  
+- shared library: C++ provides a standard API for loading shared library, and it can reduce memory usage and prevent conflicts.  
+
+In conclusion, the business needs to package the system as a shared library to run it on Ray Serve.
+### Register function
+Ray Serve will add SERVE_FUNC and SERVE_DEPLOYMENT macros to publish user Service as Deployment.  
+SERVE_FUNC: Resolving overloaded function registration;
```

Comment:
I don't fully understand the need for this part -- is it in case we have two deployments registered with the same function name?

Reply:
Yes, it is used to register overloaded methods with the same function name but different parameters.